File: bart/bart.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import (division, print_function, absolute_import,
                        unicode_literals)

__all__ = ["Star", "Planet", "PlanetarySystem"]

import numpy as np
from . import _bart
from .ld import LimbDarkening


_G = 2945.4625385377644


class Star(object):

    def __init__(self, mass=1.0, radius=1.0, flux=1.0, ldp=None):
        self.mass = mass
        self.radius = radius
        self.flux = flux

        # The limb darkening profile.
        if ldp is None:
            # Default to a uniformly illuminated star.
            self.ldp = LimbDarkening(1.0, 1.0)
        else:
            self.ldp = ldp

    def get_semimajor(self, T, planet_mass=0.0):
        """
        Get the semi-major axis for a massless planet orbiting this star with
        the period ``T``.

        """
        return (_G * T * T * (self.mass + planet_mass)
                / (4 * np.pi * np.pi)) ** (1. / 3)


class Planet(object):

    def __init__(self, r, a, mass=0.0, t0=0.0, e=0.0, pomega=0.0, ix=0.0,
                 iy=0.0):
        self.r = r
        self.a = a
        self.mass = mass
        self.t0 = t0
        self.e = e
        self.pomega = pomega
        self.ix = ix
        self.iy = iy

    def get_mstar(self, P):
        """
        Get the mass of the host star implied by the semi-major axis of this
        planet and an input period.

        """
        a = self.a
        return a * a * a * 4 * np.pi * np.pi / _G / P / P - self.mass

    def get_period(self, mstar):
        """
        Get the period of this planet orbiting a star with a given mass.

        """
        a = self.a
        return 2 * np.pi * np.sqrt(a * a * a / _G / (mstar + self.mass))


class PlanetarySystem(object):

    def __init__(self, star, iobs=90.0):
        self.star = star
        self.iobs = iobs
        self.planets = []

    @property
    def nplanets(self):
        """
        The number of planets in the system.

        """
        return len(self.planets)

    def add_planet(self, planet):
        """
        Add a :class:`Planet` to the system.

        :param planet:
            The :class:`Planet` to add.

        """
        self.planets.append(planet)

    def _get_pars(self):
        r = [(p.mass, p.r, p.a, p.t0, p.e, p.pomega, p.ix, p.iy)
             for p in self.planets]
        return zip(*r)

    def lightcurve(self, t, texp=1626.0, K=5):
        """
        Get the light curve of the model at the current model.

        :param t:
            The times where the light curve should be evaluated.

        :param texp:
            The exposure time in seconds.

        :param K:
            The number of bins to use when integrating over exposure time.

        """
        mass, r, a, t0, e, pomega, ix, iy = map(np.atleast_1d,
                                                self._get_pars())
        s = self.star
        ldp = s.ldp
        lc = _bart.lightcurve(np.atleast_1d(t), texp / 86400., K, s.flux,
                              s.mass, s.radius, self.iobs,
                              mass, r, a, t0, e, pomega, ix, iy,
                              np.atleast_1d(ldp.bins),
                              np.atleast_1d(ldp.intensity))
        return lc

File: viz/bart.js
(function () {

  "use strict";

  var bart = window.Bart = {};

  bart.LimbDarkening = function (bins, intensity) {
    // A limb darkening profile. `bins` is a list of outer bin edges and
    // `intensity` is a list of the same length with the intensity levels.
    // `bins` should be *sorted*.
    this.bins = bins.map(function (v) { return v / bins[bins.length - 1]; });

    // Normalize the intensity profile.
    var _this = this;
    var norm = Math.PI * intensity.reduce(function (n, v, i, a) {
      var r2 = _this.bins[i] * _this.bins[i];
      if (i === 0) return v * r2;
      return n + v * (r2 - _this.bins[i - 1] * _this.bins[i - 1]);
    });
    this.intensity = intensity.map(function (v) { return v / norm });
  };

  bart.occulted_area = function (R, p, b) {
    // Get the occulted area for a star of radius `R`, planet of radius `p`
    // and impact parameter `b`.
    if (b >= R + p) return 0.0;
    else if (b <= R - p) return Math.PI * p * p;
    else if (b <= p - R) return Math.PI * R * R;
    else {
      var R2 = R * R,
          p2 = p * p,
          b2 = b * b,
          k1 = Math.acos(0.5 * (b2 + p2 - R2) / b / p),
          k2 = Math.acos(0.5 * (b2 + R2 - p2) / b / R),
          k3 = Math.sqrt((p+R-b) * (b+p-R) * (b-p+R) * (b+R+p));
      return p2 * k1 + R2 * k2 - 0.5 * k3;
    }
  };

  bart.ldlc = function (p, b, ldp) {
    // Get the light curve for a planet of radius `p`, a list of impact
    // parameters `b` and a limb darkening profile.
    return b.map(function (b0) {
      var areas = ldp.bins.map(function (bin, j) {
        return bart.occulted_area(bin, p, b0);
      });
      return areas.reduce(function (result, a, j) {
        if (j === 0) return 1 - a * ldp.intensity[j];
        return result - ldp.intensity[j] * (a - areas[j - 1]);
      }, 1.0);
    });
  };

  bart.orbit = function (ts, period, a, t0, incl) {
    // Get the 3D position of a planet in a circular orbit at times `ts`. The
    // orbit has period `period`, semi-major axis `a`, initial transit time
    // `t0` and inclination `incl` (in radians).
    return ts.map(function (t) {
      var wt = 2 * Math.PI * (t - t0) / period,
          xp = a * Math.cos(wt), yp = a * Math.sin(wt);
      return [xp * Math.sin(incl), yp, xp * Math.cos(incl)];
    });
  };

  bart.Lightcurve = function (ts, period, a, t0, incl, p, ldp) {
    this.period = period;
    this.a = a;
    this.t0 = t0;
    this.incl = incl;
    this.p = p;
    this.ldp = ldp;
    this.pos = bart.orbit(ts, period, a, t0, Math.PI * incl / 180.);
    this.b = this.pos.map(function (p0) {
      if (p0[0] > 0) return Math.sqrt(p0[1]*p0[1]+p0[2]*p0[2]);
      return 1 + p;
    });
    this.time = ts;
    this.flux = bart.ldlc(p, this.b, ldp);
  };

  bart.fiducial_ldp = function (g1, g2) {
    // Get a quadratic limb darkening profile.
    var bins = [],
        intensity = [];
    for (var i = 1; i <= 25; ++i) {
      var r = i / 25.,
          onemmu = 1 - Math.sqrt(1 - r * r);
      bins.push(r);
      intensity.push(1 - g1 * onemmu - g2 * onemmu * onemmu);
    }
    return new bart.LimbDarkening(bins, intensity);
  };

})();
